### PR TITLE
add basic AbstractInterpreter

### PR DESCRIPTION
**Pull request by vchuravy**

### CODE REVIEW

File: src/compiler/interpreter.jl [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+else # @static if isdefined(Compiler, :is_stmt_inline)
+
+import Core.Compiler: InliningTodo, InliningState
+enzyme_inlining_policy(src) = Core.Compiler.default_inlining_policy(src)
+Core.Compiler.inlining_policy(::EnzymeInterpeter) = enzyme_inlining_policy
+function Core.Compiler.resolve_todo(todo::InliningTodo, state::InliningState{S, T, <:typeof(enzyme_inlining_policy)}) where {S<:Union{Nothing, Core.Compiler.EdgeTracker}, T}
+    mi = todo.mi
+    @show mi.specTypes
+    @show is_primitive_func(mi.specTypes)
+    if is_primitive_func(mi.specTypes)
```
Comment (aviatesk) [inline src/compiler/interpreter.jl:101]: `autodiff(foo, Active(2.0))` certainly hits this case.
It seems like you need to fix another part of Enzyme as well?
```julia
julia> autodiff(foo, Active(2.0)) # ERROR: LLVM error: unknown binary operator, see below for details
mi.specTypes = Tuple{typeof(log), Float64}
ERROR: MethodError: no method matching classify_arguments(::GPUCompiler.CompilerJob{Enzyme.Compiler.EnzymeTarget, Enzyme.Compiler.EnzymeCompilerParams, GPUCompiler.FunctionSpec{typeof(foo), Tuple{Float64}}}, ::LLVM.FunctionType)
Closest candidates are:
  classify_arguments(::GPUCompiler.CompilerJob, ::LLVM.Function) at ~/.julia/packages/GPUCompiler/fG3xK/src/irgen.jl:294
Stacktrace:
 [1] lower_convention(job::GPUCompiler.CompilerJob, mod::LLVM.Module, entry_f::LLVM.Function, actualRetType::Type)
   @ Enzyme.Compiler ~/julia/packages/Enzyme/src/compiler.jl:2007
 [2] codegen(output::Symbol, job::GPUCompiler.CompilerJob{Enzyme.Compiler.EnzymeTarget, Enzyme.Compiler.EnzymeCompilerParams, GPUCompiler.FunctionSpec{typeof(foo), Tuple{Float64}}}; libraries::Bool, deferred_codegen::Bool, optimize::Bool, strip::Bool, validate::Bool, only_entry::Bool, parent_job::Nothing)
   @ Enzyme.Compiler ~/julia/packages/Enzyme/src/compiler.jl:2273
 [3] _thunk(job::GPUCompiler.CompilerJob{Enzyme.Compiler.EnzymeTarget, Enzyme.Compiler.EnzymeCompilerParams, GPUCompiler.FunctionSpec{typeof(foo), Tuple{Float64}}})
   @ Enzyme.Compiler ~/julia/packages/Enzyme/src/compiler.jl:2628
 [4] cached_compilation(cache::Dict{UInt64, Any}, job::GPUCompiler.CompilerJob, compiler::typeof(Enzyme.Compiler._thunk), linker::typeof(Enzyme.Compiler._link))
   @ GPUCompiler ~/.julia/packages/GPUCompiler/fG3xK/src/cache.jl:89
 [5] thunk(f::typeof(foo), df::Nothing, ::Type{Active{Float64}}, tt::Type{Tuple{Active{Float64}}}, ::Val{false})
   @ Enzyme.Compiler ~/julia/packages/Enzyme/src/compiler.jl:2680
 [6] autodiff
   @ ~/julia/packages/Enzyme/src/Enzyme.jl:186 [inlined]
 [7] autodiff(f::typeof(foo), args::Active{Float64})
   @ Enzyme ~/julia/packages/Enzyme/src/Enzyme.jl:216
 [8] top-level scope
   @ REPL[3]:1
```

Comment (vchuravy) [reply src/compiler/interpreter.jl:101]: Which version of Julia are you on? I am testing with 1.7 and you seem to have gotten an old version of GPUCompiler.

Comment (vchuravy) [reply src/compiler/interpreter.jl:105]: You are right... It does work. Yay!

File: src/compiler/interpreter.jl [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+function is_primitive_func(@nospecialize(TT))
+    ft = TT.parameters[1]
+    if ft === typeof(Base.cbrt) || ft === typeof(Base.sin) || ft === typeof(Base.cos) ||
+       ft === typeof(Base.tan) || ft === typeof(Base.exp) || ft === typeof(Base.log) ||
+       ft === typeof(Base.asin) || ft === typeof(Base.tanh) || ft === typeof(Base.FastMath.tanh_fast)
+
+        if TT <: Tuple{ft, Float32} || TT <: Tuple{ft, Float64}
+            return true
+        end
+    end
```
Comment (wsmoses) [inline src/compiler/interpreter.jl:75]: Add wait and enq_task

File: src/compiler/interpreter.jl (hunk 1 above)
Comment (aviatesk) [reply src/compiler/interpreter.jl:104]: Yeah, after updating deps, now it works:)

File: src/compiler/interpreter.jl [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+# No need to do any locking since we're not putting our results into the runtime cache
+Core.Compiler.lock_mi_inference(interp::EnzymeInterpeter, mi::MethodInstance) = nothing
+Core.Compiler.unlock_mi_inference(interp::EnzymeInterpeter, mi::MethodInstance) = nothing
+
+function Core.Compiler.add_remark!(interp::EnzymeInterpeter, sv::InferenceState, msg)
+end
+
+Core.Compiler.may_optimize(interp::EnzymeInterpeter) = true
+Core.Compiler.may_compress(interp::EnzymeInterpeter) = true
+Core.Compiler.may_discard_trees(interp::EnzymeInterpeter) = true
```
Comment (aviatesk) [inline src/compiler/interpreter.jl:53]: `may_discard_trees = true` means a complicated (in terms of inlineability) source will be discarded, but as far as I understand Enzyme wants "always inlining, except special cased functions", so I guess we really don't want to discard sources?

File: src/compiler/interpreter.jl [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+if isdefined(Base.Experimental, Symbol("@overlay"))
+Core.Compiler.method_table(interp::EnzymeInterpeter, sv::InferenceState) =
+    Core.Compiler.OverlayMethodTable(interp.world, interp.method_table)
+else
+Core.Compiler.method_table(interp::EnzymeInterpeter, sv::InferenceState) =
+    GPUCompiler.WorldOverlayMethodTable(interp.world)
+end
+
+function is_primitive_func(@nospecialize(TT))
+    ft = TT.parameters[1]
```
Comment (aviatesk) [inline src/compiler/interpreter.jl:67]:
```suggestion
    isa(TT, DataType) || return false
    ft = TT.parameters[1]
```

File: src/compiler/interpreter.jl [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    mi::MethodInstance, argtypes::Vector{Any})
+
+    if is_primitive_func(mi.specTypes)
+        return nothing
+    end
+
+    return Base.@invoke Core.Compiler.inlining_policy(
+        interp::AbstractInterpreter, nothing, stmt_flag::UInt8,
+        mi::MethodInstance, argtypes::Vector{Any})
+    return nothing
```
Comment (aviatesk) [inline src/compiler/interpreter.jl:94]:
```suggestion
```

File: src/compiler/interpreter.jl (hunk 3 above)
Comment (vchuravy) [reply src/compiler/interpreter.jl:53]: We don't need to always inline everyting Enzyme will differentiate through invoke and generic calls.

File: src/compiler/interpreter.jl [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+Core.Compiler.method_table(interp::EnzymeInterpeter, sv::InferenceState) =
+    Core.Compiler.OverlayMethodTable(interp.world, interp.method_table)
+else
+Core.Compiler.method_table(interp::EnzymeInterpeter, sv::InferenceState) =
+    GPUCompiler.WorldOverlayMethodTable(interp.world)
+end
+
+function is_primitive_func(@nospecialize(TT))
+    isa(TT, DataType) || return false
+    isa(TT, DataType) || return false
```
Comment (vchuravy) [inline src/compiler/interpreter.jl:68]:
```suggestion
```

File: src/compiler/interpreter.jl (hunk 3 above)
Comment (wsmoses) [reply src/compiler/interpreter.jl:53]: I mean it's still more efficient if those invoke/generic can be deduced

File: src/compiler/interpreter.jl [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+function Core.Compiler.inlining_policy(
+    interp::EnzymeInterpeter, @nospecialize(src), stmt_flag::UInt8,
+    mi::MethodInstance, argtypes::Vector{Any})
+
+    if is_primitive_func(mi.specTypes)
+        return nothing
+    end
+
+    return Base.@invoke Core.Compiler.inlining_policy(
+        interp::AbstractInterpreter, nothing, stmt_flag::UInt8,
```
Comment (aviatesk) [inline src/compiler/interpreter.jl:93]:
```suggestion
        interp::AbstractInterpreter, src::Any, stmt_flag::UInt8,
```
`src === nothing` here means we may inline only those calls with callsite `@inline` annotations.

File: src/compiler/interpreter.jl [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+        mi::MethodInstance, argtypes::Vector{Any})
+end
+
+elseif isdefined(Core.Compiler, :inlining_policy)
+
+import Core.Compiler: InliningTodo, InliningState
+enzyme_inlining_policy(src) = Core.Compiler.default_inlining_policy(src)
+Core.Compiler.inlining_policy(::EnzymeInterpeter) = enzyme_inlining_policy
```
Comment (aviatesk) [inline src/compiler/interpreter.jl:101]:
```suggestion
Core.Compiler.inlining_policy(::EnzymeInterpeter) = Core.Compiler.default_inlining_policy
```
or better to add `@nospecialize` annotation to `enzyme_inlining_policy`.

File: src/compiler/interpreter.jl (hunk 3 above)
Comment (aviatesk) [reply src/compiler/interpreter.jl:53]: I didn't know that GPUCompiler already handles generic (dynamic) calls that well.

Comment (vchuravy) [reply src/compiler/interpreter.jl:53]: It doesn't but Enzyme rewrites them to it's own handler that re-enters the Enzyme compiler

File: src/compiler/interpreter.jl (hunk 8 above)
Comment (vchuravy) [reply src/compiler/interpreter.jl:101]: Yeah we are dispatching one `enzyme_inlining_policy` in `resolve_todo`

File: src/compiler/interpreter.jl [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+    return Base.@invoke Core.Compiler.inlining_policy(
+        interp::AbstractInterpreter, nothing, stmt_flag::UInt8,
+        mi::MethodInstance, argtypes::Vector{Any})
+end
+
+elseif isdefined(Core.Compiler, :inlining_policy)
+
+import Core.Compiler: InliningTodo, InliningState
+enzyme_inlining_policy(src) = Core.Compiler.default_inlining_policy(src)
```
Comment (vchuravy) [inline src/compiler/interpreter.jl:100]:
```suggestion
enzyme_inlining_policy(@nospecialize(src)) = Core.Compiler.default_inlining_policy(src)
```

**Comment by vchuravy** [comment]
I am worried that this will run into https://github.com/JuliaLabs/Cassette.jl/issues/195 for the GPU support as well as KA...

**Comment by aviatesk** [comment]
Where is Cassette used within Enzyme's execution pipeline?

**Comment by vchuravy** [comment]
Cassette isn't used anymore, but we will run two abstract interpreters. One for CUDA.jl on the outside and then one for Enzyme...

**Pull request merged**